### ISSUE TITLE
Initiating a Direct transfer gets stuck when partner is offline

### ISSUE THREAD

**Issue by LefterisJP** [raiden-network/raiden#900]
## Problem Definition

If we have a channel with a partner and attempt a (Direct) transfer while he is offline then the API call will get stuck.

## Drawbacks

The counterparty may be malicious, and actually just withhold the ACK, while he already has sent us the transfer.

## Solution

This needs thinking:

End the API call after a timeout and provide an error message to the user if we don't get the `ACK` on time. But we have to find a way to handle the drawback of a malicious partner pretending to be offline.

The most direct way to handle this would be to end the attempted transfer after the timeout because we have not received an ACK, provide an error message to the user and then immediately close the channel and go in the blockchain since any transfers from that point could be unsafe.

**Comment by hackaugusto** [comment]
I don't understand which api is problematic, is it the rest or python api?

Important details:
- A user would expect a timeout to entail in transfer cancellation, and *this is not the case for the moment*, meaning that the transfer message will stay in the queue until it's acknowledged and the task will stay alive until the lock expire. A cancel timeout must dispatch an `ActionCancelTransfer` state change, *this state change may fail*. This is quite important since right now a timeout does not mean the transfer failed or succeed, only that the sender gave up waiting for a result.
- Only a mediated transfer may be canceled, it doesn't matter if an ack is not received for a direct transfer, if the message is sent the transfer should be considered complete. IOW direct transfers can not be cancelled.
- A cancelled mediated transfer does not mean the message queue should be touched, only that the secret will be discarded and the withdraw of the lock is impossible (given that sha3 has good collision resistance)

**Comment by LefterisJP** [comment]
Right now this issue is only about simple Direct Transfers.

> Only a mediated transfer may be canceled, it doesn't matter if an ack is not received for a direct transfer, if the message is sent the transfer should be considered complete. IOW direct transfers can not be cancelled.

When we [initiate a transfer](https://github.com/raiden-network/raiden/blob/master/raiden/api/rest.py#L391) which will end up as a direct transfer via the API the call never returns. I assumed this is because we somehow wait for an ACK, but I am not sure yet.

Which is funny, because if you follow the code down the line it should all by asynchronous and the call should return immediately.

Looking at some `log.DEV()` calls I put down the code path it seems the call proceeds fine until the end of the protocol:

```
CRITICAL:raiden.api.v1.resources        Inside Transfer post endpoint handler                                                           
CRITICAL:raiden.api.rest        Inside rest.initiate_transfer                                                                           
CRITICAL:raiden.api.python      Inside api.python.transfer_async                                                                        
CRITICAL:raiden.raiden_service  We are making a direct transfer                                                                         
CRITICAL:raiden.raiden_service  Right before protocol send                                                                              
DEBUG:raiden.network.protocol   new queue created for (9bed7fd1, 0f114a1e) > 47314ee7                                                                                                                                                                                            
CRITICAL:raiden.network.protocol        Sent message with ecohash echohash=189b4218                                                                                                                                                                                              
```

Will need to look further into this issue.

**Comment by LefterisJP** [comment]
So as suspected we send the transfer and then [wait on it](https://github.com/raiden-network/raiden/blob/master/raiden/api/python.py#L480), and if the other party is offline and we don't get an ACK we end up waiting forever. There should be a timeout there and if it hits we should close the channel for safety. What do you think @hackaugusto?

Or another option is to switch to the simple `transfer_async` for direct transfers and once sent, always consider it as received by the other side. But I think the timeout with the ACK makes a bit more sense.

**Comment by andrevmatos** [comment]
@LefterisJP I have something like that on that testing branch on #895 . I may clean it and make a new PR, so at least, we don't wait and hang the request forever. Should I do it?

**Comment by hackaugusto** [comment]
> There should be a timeout there and if it hits we should close the channel for safety
I don't think closing the channel is necessary, it does not increase safety since the waiting node is not a risk of losing token. If there is a transfer with risky of losing tokens the mediated task will close it.

Also, since this is a direct transfer, a cancel action doesnt make sense.

I think the caller should provide a timeout that it thinks is good enough, for the REST api that means the caller will close the connection in which the HTTP request is being done.

**Comment by LefterisJP** [comment]
> I think the caller should provide a timeout that it thinks is good enough, for the REST api that means the caller will close the connection in which the HTTP request is being done.

But we have no way to do so from the Web API for example. Would it not be better to just provide a proper default timeout value in the `async_result.wait()` ?

**Comment by hackaugusto** [comment]
> But we have no way to do so from the Web API for example

Why not? Why closing the http request is not good enough?

> Would it not be better to just provide a proper default timeout value in the async_result.wait() ?

I don't think so, timeouts don't compose well.

**Comment by LefterisJP** [comment]
> Why not? Why closing the http request is not good enough?

I actually am not sure. For the Webui we have to ask @andrevmatos 

> I don't think so, timeouts don't compose well.

I really don't see this as a nice solution usability wise. Compare these 2 user scenarios.

### Scenario 1

Alice starts a transfer via curl. She gets no ACK from BoB after `X` seconds and the timeout kicks in, stopping the transfer attempt, finishing the http request and informing her that the transfer timed out and that BoB is probably offline.

### Scenario 2

Alice starts a transfer via curl. She gets no ACK from BoB after `X` seconds. She keeps waiting. At some point she will start wondering what the heck is wrong, will have to read the docs and find out that this is expected behaviour if BoB is offline and that she should take action to close the http request manually.
### Clarification

Scenario 1 is a thousand times better for our users.

What do you mean by "compose well".

When we have the direct transfers as parts of a mediated transfer and we have more than one timeouts to consider? If so we can specify the timeout only for simple direct transfers.

**Comment by andrevmatos** [comment]
I, too, think putting a timeout on async_result is nice, and maybe even changing `wait` to `get`, so we can have exceptions raised on the request processing greenlet, if they happen down the line.

**Comment by hackaugusto** [comment]
I don't agree with these two scenarios, the user should read the docs and errors should be handle anyways, if you want to convey the information that the operation is blocking, the URL name may be changed to say blocking, e.g. `api/1/blocking_transfer/`

> What do you mean by "compose well".

Timeouts are not shared information among the different layers of the system, it's hard to synchronize them. The inner layers of the system must always have a total time larger than the outer layers *without* knowing what are the outer layer timeout specs. The easiest way of fixing this is pushing the choice to the edges, to the user.

**Comment by LefterisJP** [comment]
@hackaugusto I see your point but I don't necessarily agree with the absence of a timeout in the api's transfers.

 Let's see what @heikoheiko also thinks about this. Either way a change needs to happen, at least in the docs, because right now it really feels wrong to use as is.

**Comment by LefterisJP** [comment]
Another thing to note is that if the timeout does occur and the transfer is considered failed then we still modify our `transferred_amount` in the channel.

Is this desired behaviour?

For example, if I try to send 1 token to Andre 3 times and he is offline or his uPnP is not working and is unreachable I will still have my balance set to `Deposit - 3` and not `Deposit - 1`. Once he fixes it and comes back online if I try to send him the transfer that was failing, so an 1 token transfer, and if he receives it he will receive `4` tokens and my balance will be `Deposit - 4`

In some scenarios this would be desirable behaviour, but would it always make sense? What if we have a direct channel with someone and we send him 1 token to buy a real life product and he says he did not receive it due to network errors? 

We can't retry. Retrying would mean that we would send him 2 times the amount.

So the question here is:

Should the transferred amount be modified once the direct transfer is sent (current behaviour) or only once the ACK is received?

**Comment by andrevmatos** [comment]
I can see from the logs that the 4 transfers have the same `identifier`, and even so, they're somehow accounted as different transfers. Maybe we could enforce the `identifier` uniqueness (and have it signed too, somehow) to evaluate (on both sides) the value increment/signing?

**Comment by LefterisJP** [comment]
No the identifier is only something to identify the transfer for third party apps, it is not a nonce and should not be enforced anywhere in the protocol.

**Comment by hackaugusto** [comment]
> Another thing to note is that if the timeout does occur and the transfer is considered failed then we still modify our transferred_amount in the channel.
> Is this desired behaviour?

Yes, Direct transfer are **not cancelable**, and they do have an inherited level of trust among the payer and payee, if someone wants to have a cancelable transfer a mediated transfer must be used. Also note that mediated transfers atm have the same problem, but with the Secret message instead of the the MediatedTransfer message. 

> We can't retry. Retrying would mean that we would send him 2 times the amount.

Just to make this point clear, Raiden can and does retry sending messages, the messages are enqueued in-order and they have a monotonically increasing transferred amount for this reason, iow the messages are idempotent and can be sent multiple times. The user cannot call the transfer method twice because the interface exposed to the user is not idempotent, if an idempotent interface should be expose we could require the user to send the `transferred_amount` instead of the the *transfer amount*, another option is to use the identifiers to check if the transfer is already in flight.

> Should the transferred amount be modified once the direct transfer is sent (current behaviour) or only once the ACK is received?

Modifying the transferred amount on the ACK breaks the protocol and makes no sense. This would mean that the direct transfer should not be a balance proof anymore, and that an additional message that is a balance proof should be sent once the ACK is received, thus only adding an additional round-trip without changing anything, since the payee can claim the balance proof was not received (in this the case the hypothetical second message). If you want to remove trust check the receipt issue https://github.com/raiden-network/raiden/issues/368

> Maybe we could enforce the identifier uniqueness (and have it signed too, somehow) to evaluate (on both sides) the value increment/signing?

I agree with this approach.

**Comment by LefterisJP** [comment]
> Just to make this point clear, Raiden can and does retry sending messages, the messages are enqueued in-order and they have a monotonically increasing transferred amount for this reason, iow the messages are idempotent and can be sent multiple times.

Ofcourse, but the whole point is that if the user can't have an idempotent transfer interface then "raiden transfers" are not idempotent.  At this point with the release so close we do care about what the user can and can not do and we do define raiden by it.

## Solution

> The user cannot call the transfer method twice because the interface exposed to the user is not idempotent, if an idempotent interface should be expose we could require the user to send the transferred_amount instead of the the transfer amount, another option is to use the identifiers to check if the transfer is already in flight.

> Maybe we could enforce the identifier uniqueness (and have it signed too, somehow) to evaluate (on both sides) the value increment/signing?
> I agree with this approach.

We need to provide the user with the ability to create idempotent direct transfers. We can't stay with the current model. So it seems we have two choices then.

### Use the identifier

So far we were regarding the identifer only as something to be used by applications built on top of raiden and not as a part of our protocol. The original [issue](https://github.com/raiden-network/raiden/issues/139) left its purpose of existence a bit vague.

How would it work?

We could have a map of identifiers to direct transfers in flight and if another user request comes for a direct transfer that is in flight and in the map we simply ignore it.

Anything more?

### Expose the transferred amount

Agreeing on the transferred amount and exposing that interface to the user to have the ability to do idempotent transfers is also possible. The problem with this approach is that it does not sound very intuitive to the user. How would it work?

I have already sent 4 tokens to Bob. I now want to send another 2 tokens to Bob so I am going to sign a message with 6 being the transferred amount. The problem is a user interface problem. How to show to the user that these 2 tokens he is signing to send are gonna bring his total balance to `deposit - 6` and not to `deposit - 8` or `deposit - 10` if for some reason he tries to repeat the transfer.

At the moment I am inclined to say we should go for the identifier approach.

**Comment by andrevmatos** [comment]
Yes, from my perspective, exposing/requiring this idempotent interface to the user adds an unecessary burden on users, and increase complexity from user's perspective/interface building. The identifier is already there, we wouldn't even need to change user's exposed APIs.
I would add that, besides ignoring requests with already in flight message sending, we should keep some registry so, for timed-out (or other somehow failed message sending, e.g.: no ACK received) and same ID request, we should go for the same final balance signing of previous request (understanding that `same id == same balance`), instead of re-incrementing and signing a new higher balance, and, oc, sending again this message. Users requesting a new transfer with different balance and same `identifier` is an error.

**Comment by hackaugusto** [comment]
> We need to provide the user with the ability to create idempotent direct transfers. We can't stay with the current model. So it seems we have two choices then.

Just to emphasize my opinion, all APIs when possible should be idempotent, otherwise retries are not safe, the API for depositing is one that is also problematic.

> So far we were regarding the identifer only as something to be used by applications built on top of raiden and not as a part of our protocol. The original issue left its purpose of existence a bit vague.
> How would it work?

The protocols stay the same, neither the smart contract nor the network interactions need to change. This only need to affect the local node data.

> We could have a map of identifiers to direct transfers in flight and if another user request comes for a direct transfer that is in flight and in the map we simply ignore it.

I don't think we should just ignore it, for the REST API the server should hold the request until the transfer completes or the lock expires. This allows the client to restart.

Also the map should be local and only for the outgoing transfers, shared among direct and mediated transfers. As "local and for outgoing" I mean that transfers were the node is the mediator should not end up in the map, nor transfer that the node is the target. (not sure about the exchange tasks)

> Agreeing on the transferred amount and exposing that interface to the user to have the ability to do idempotent transfers is also possible. The problem with this approach is that it does not sound very intuitive to the user. How would it work?

Well, the client would need to track the latest transferred amount, and transferred amount with a lower value than the latest balance proof would just be ignored. I don't think it's a necessarily a good interface, just pointed it out as an alternative. 
  
@andrevmatos could you rephrase the second paragraph? I didn't understand you.

**Comment by andrevmatos** [comment]
Yes, @hackaugusto , sorry for that, @LefterisJP also complained it was not clear.
The idea is to keep this mapping not only for in-flight requests, but at least until the next request of that type (e.g. in transfer case, until another transfer is made on the same channel), or even the end of the session. That way, if the request failed (timed out, no ACK received, network issues, etc), the user can retry the same request, with same `identifier` and same payload (in transfer case, the amount to be transfered), and Raiden will handle it as being idempotent, signing the same `transfered_amount`, instead of making a new, possible doubled, transfer (`identifier` being the unique key to identify requests). Making a new request with same `identifier` and different payload is an error. This same approach could be used in the future for every state-changing request, avoiding double spends and allowing safe retries by clients.
